### DISCUSSION TITLE
Plans for 2.x

### DISCUSSION THREAD

**Discussion by ahx** [ahx/openapi_first#262] [General]
Hi. Here is what I am working on for 2.x:

UPDATE: Splitted in 2.0 and beyond

## 2.0
1. ✅ A more structured code base. – Currently some things are grouped pretty randomly.
1. ✅ Performance improvements. – Faster is better.
1. ✅ Hooks! – Run code after validation was called. See https://github.com/ahx/openapi_first/tree/dev?tab=readme-ov-file#hooks
1. ✅ Assertion methods to do contract testing without using the middlewares. 
1. ✅ [Getting the full response with the ResponseInvalidError](https://github.com/ahx/openapi_first/issues/255)

## 2.x
1. An ambitious (test-)coverage implementation. – Coverage should, for example, report about un-tested query parameters.
1. [Validating OpenAPI files](https://github.com/ahx/openapi_first/issues/259)?

**Comment by MrBananaLord** [comment]
Thanks for putting it all together in a plan! 🌟 

Do you have an idea on how to approach the assertions? From the looks of it something like this should be enough for the request specs:
```
let(:definition) { OpenapiFirst.load("./openapi.yml") }

it "fails when the 'bar' param is not an integer" do
  get "/foo", params: { bar: "asd" } 

  openapi_first_request = definition.request(Rack::Request.new(request.env))
  
  expect(openapi_first_request.valid?).to eq(false)
end

it "fails when the 'foo' in response is not an integer" do
  get "/foo", params: { bar: "asd" } 

  openapi_first_request = definition.request(Rack::Request.new(request.env))
  openapi_first_response = openapi_first_request.response(Rack::Response[*response.to_a])
  
   expect(openapi_first_response.valid?).to eq(false)
end
```
maybe extracting the logic above to separate assertions would be enough? 🤔

**Reply by ahx** [reply]
@MrBananaLord Yes, I think that should be enough. Having something a little more concise without the need to instantiate Rack::Request etc would be nice.

```rspec
it "fails when the 'bar' param is not an integer" do
  get "/foo", params: { bar: "asd" } 

  expect(request).to be_valid_request(super_api_description)
end
```

or 

```rspec
it "fails when the 'bar' param is not an integer" do
  get "/foo", params: { bar: "asd" } 

  assert_something_something(request, definition) # I cannot think of a good name
end
```

[openapi_contracts](https://github.com/mkon/openapi_contracts/tree/main?tab=readme-ov-file#usage) and [skooma](https://github.com/skryukov/skooma?tab=readme-ov-file#validate-request) have closer integration with RSpec, but I could also think of having a very simple `assert_something` method that just raises an exception with a useful error message and having no test-framework specifics going on, similar to [committee](https://github.com/interagent/committee?tab=readme-ov-file#test-assertions).
I would love to get some feedback about what people would like to use here.

**Comment by ahx** [comment]
Hi. I have just merged my local changes into main and updated the plan above. I would like to release 2.0 soon and focus on test coverage and the others later.

The biggest change on main is probably the `assert_api_conform` method ([see Readme](https://github.com/ahx/openapi_first?tab=readme-ov-file#test-assertions)), which makes openapi_first more usable in situations where you don't want to use the middlewares.

Any feedback would be very much appreciated.

**Comment by ahx** [comment]
2.0 has just been release ([CHANGELOG](https://github.com/ahx/openapi_first/blob/main/CHANGELOG.md#20))
